xinference/model/llm/llm_family.py

```diff
@@ -132,3 +153,125 @@ def cache_from_huggingface(
+) -> Optional[Tuple[LLMFamilyV1, LLMSpecV1, str]]:
+    """
+    Find an LLM family, spec, and quantization that satisfy given criteria.
+    """
+    user_defined_llm_families = get_user_defined_llm_families()
+
+    for family in BUILTIN_LLM_FAMILIES + user_defined_llm_families:
+        if model_name != family.model_name:
+            continue
+        for spec in family.model_specs:
+            if (
+                model_format
+                and model_format != spec.model_format
+                or model_size_in_billions
+                and model_size_in_billions != spec.model_size_in_billions
+                or quantization
+                and quantization not in spec.quantizations
+            ):
+                continue
+            if quantization:
+                return family, spec, quantization
+            else:
+                # by default, choose the most coarse-grained quantization.
+                # TODO: too hacky.
+                quantizations = spec.quantizations
+                quantizations.sort()
+                for q in quantizations:
+                    if (
+                        is_local_deployment
+                        and not (_is_linux() and _has_cuda_device())
+                        and q == "4-bit"
+                    ):
+                        logger.warning(
+                            "Skipping %s for non-linux or non-cuda local deployment .",
+                            q,
+                        )
+                        continue
+                    return family, spec, q
+    return None
+
+
+def register_llm(llm_family: LLMFamilyV1, persist: bool):
+    from .utils import is_valid_model_name
+
+    if not is_valid_model_name(llm_family.model_name):
+        raise ValueError(
+            f"Invalid model name {llm_family.model_name}. The model name must start with a letter"
+            f" or a digit, and can only contain letters, digits, underscores, or dashes."
+        )
+
+    with UD_LLM_FAMILIES_LOCK:
+        for family in BUILTIN_LLM_FAMILIES + UD_LLM_FAMILIES:
+            if llm_family.model_name == family.model_name:
+                raise ValueError(
+                    f"Model name conflicts with existing model {family.model_name}"
+                )
+
+        UD_LLM_FAMILIES.append(llm_family)
+
+    if persist:
+        persist_path = os.path.join(
+            XINFERENCE_MODEL_DIR, "llm", f"{llm_family.model_name}.json"
+        )
+        os.makedirs(os.path.dirname(persist_path), exist_ok=True)
+        with open(persist_path, mode="w") as fd:
+            fd.write(llm_family.json())
+
+
+def unregister_llm(model_name: str, permanent: bool):
+    with UD_LLM_FAMILIES_LOCK:
+        llm_family = None
+        for i, f in enumerate(UD_LLM_FAMILIES):
+            if f.model_name == model_name:
+                llm_family = f
+                break
+        if llm_family:
+            UD_LLM_FAMILIES.remove(llm_family)
+            if permanent:
```

Comment:
The word `permanent` is ambiguous. When `permanent=True`, the file path is deleted. The semantics do not match.

Reply:
Removed.